rasterio/merge.py

```diff
@@ -340,46 +340,33 @@ def nullcontext(obj):
             if first_crs != src.crs:
                 raise RasterioError(f"CRS mismatch with source: {dataset}")
 
-            # 1. Compute spatial intersection of destination and source
-            src_w, src_s, src_e, src_n = src.bounds
+            # 1. Compute the source window
+            src_window = windows.from_bounds(dst_w, dst_s, dst_e, dst_n, src.transform)
 
-            int_w = src_w if src_w > dst_w else dst_w
-            int_s = src_s if src_s > dst_s else dst_s
-            int_e = src_e if src_e < dst_e else dst_e
-            int_n = src_n if src_n < dst_n else dst_n
-
-            # 2. Compute the source window
-            src_window = windows.from_bounds(int_w, int_s, int_e, int_n, src.transform)
-
-            # 3. Compute the destination window
+            # 2. Compute the destination window
             dst_window = windows.from_bounds(
-                int_w, int_s, int_e, int_n, output_transform
+                dst_w, dst_s, dst_e, dst_n, output_transform
             )
 
-            # 4. Read data in source window into temp
-            src_window_rnd_shp = src_window.round_lengths()
-            dst_window_rnd_shp = dst_window.round_lengths()
-            dst_window_rnd_off = dst_window_rnd_shp.round_offsets()
```

Comment:
All of this rounding seems to be superfluous. Removing it doesn't break any tests, which gives me a bad feeling. We definitely need more tests for merging.